ledger-service/http-json/src/main/scala/com/digitalasset/http/Main.scala

```diff
@@ -50,6 +55,35 @@ object Main {
     config.logLevel.foreach(GlobalLogLevel.set("Ledger HTTP-JSON API"))
   }
 
+  def executeWithShutdownHookNet[F[_], A](
+      program: Deferred[F, Unit] => F[A],
+      shutdownHookTimeout: FiniteDuration = 10.seconds,
+  )(implicit
+      F: Applicative[F],
+      globalConcurrentEffect: ConcurrentEffect[F],
+      globalTimer: Timer[F],
+      loggingCtx: LoggingContextOf[InstanceUUID],
+  ): F[A] = {
+    import cats.FlatMap.ops.toAllFlatMapOps
+    for {
+      _ <- F.pure(()): F[Unit]
+      shutdownHookCalled <- Deferred[F, Unit](globalConcurrentEffect)
+      didProgramAlreadyEnd <- Deferred[F, Unit](globalConcurrentEffect)
+      _ = sys.addShutdownHook {
+        globalConcurrentEffect.toIO(shutdownHookCalled.complete(())).unsafeRunSync()
+
+        logger.debug(
+          s"Shutdown hook will wait $shutdownHookTimeout for the application to exit gracefully otherwise it will forcefully shutdown"
+        )
+        globalConcurrentEffect
+          .toIO(Concurrent.timeoutTo(didProgramAlreadyEnd.get, shutdownHookTimeout, F.unit))
+          .unsafeRunSync()
+      }
+      result <- program(shutdownHookCalled)
+      _ <- didProgramAlreadyEnd.complete(())
+    } yield result
+  }
```

Comment:
Wouldn't this be simpler without Cats Effect? I'm far from expert in this lib, but it seems extremely hard to follow what's going on

Reply:
I don't think this is going to get that much simpler if I get rid of cats effect here.

Reply:
You might be right, I am not an expert as I mentioned.